ark/phenotyping/som_utils.py

```diff
@@ -131,50 +111,174 @@ def compute_pixel_cluster_channel_avg(fovs, channels, base_dir, cluster_col,
-                                     fovs=None, cluster_col='cluster'):
-    """Compute the average marker expression for each cell
+def compute_cell_cluster_channel_avg(fovs, channels, base_dir,
+                                     weighted_cell_channel_avg_name,
+                                     cell_cluster_name='cell_mat_clustered.feather',
+                                     cell_cluster_col='cell_meta_cluster'):
+    """Computes the average marker expression for each cell cluster
+
+    Args:
+        fovs (list):
+            The list of fovs to subset on
+        channels (list):
+            The list of channels to subset on
+        base_dir (str):
+            The path to the data directory
+        weighted_cell_channel_avg_name (str):
+            The name of the weighted cell table, created in `example_cell_clustering.ipynb`
+        cell_cluster_name (str):
+            Name of the file containing the cell data with cluster labels
+        cell_cluster_col (str):
+            Whether to aggregate by cell SOM or meta labels
+            Needs to be either 'cell_som_cluster', or 'cell_meta_cluster'
+
+    Returns:
+        pandas.DataFrame:
+            Each cell cluster mapped to the average expression for each marker
+    """
+
+    # verify the cell table actually exists
+    if not os.path.exists(os.path.join(base_dir, weighted_cell_channel_avg_name)):
+        raise FileNotFoundError(
+            "Weighted cell table %s not found in %s" % (weighted_cell_channel_avg_name, base_dir)
+        )
+
+    # verify the cell cluster col specified is valid
+    misc_utils.verify_in_list(
+        provided_cluster_col=[cell_cluster_col],
+        valid_cluster_cols=['cell_som_cluster', 'cell_meta_cluster']
+    )
+
+    cell_table = pd.read_csv(os.path.join(base_dir, weighted_cell_channel_avg_name))
+
+    # subset on only the fovs the user has specified
+    cell_table = cell_table[cell_table['fov'].isin(fovs)]
+
+    # read the clustered data
+    cluster_data = feather.read_dataframe(os.path.join(base_dir, cell_cluster_name))
+
+    # need to ensure that both cell_table and cluster_data have FOVs and segmentation_labels sorted
+    # in the same order, this can be done by simply sorting by fov and segmentation_label for both
+    cell_table = cell_table.sort_values(
+        by=['fov', 'segmentation_label']
+    ).reset_index(drop=True)
+    cluster_data = cluster_data.sort_values(
+        by=['fov', 'segmentation_label']
+    ).reset_index(drop=True)
+
+    # add an extra check to ensure that the FOVs and segmentation labels are in the same order
+    misc_utils.verify_same_elements(
+        enforce_order=True,
+        cell_table_fovs=list(cell_table['fov']),
+        cluster_data_fovs=list(cell_table['fov'])
+    )
+    misc_utils.verify_same_elements(
+        enforce_order=True,
+        cell_table_labels=list(cell_table['segmentation_label']),
+        cluster_data_labels=list(cell_table['segmentation_label'])
```

Comment:
I think you meant cluster_data_fovs=list(cluster_data['segmentation_label']) here?